### PR TITLE
Document how to enable/disable Debug Output on the fly

### PR DESCRIPTION
**Pull request by Al2Klimov**
This is a good alternative to `icinga2 feature enable debuglog`:

* Object creation/deletion via API happens immediately and requires no restart
* Hence, the debug log is enabled exactly as long as desired

### CODE REVIEW

File: doc/15-troubleshooting.md [hunk 1]
```diff
@@ -176,6 +176,60 @@ C:\> cd C:\ProgramData\icinga2\var\log\icinga2
 C:\ProgramData\icinga2\var\log\icinga2> Get-Content .\debug.log -tail 10 -wait
 ```
 
+### Enable/Disable Debug Output on the fly <a id="troubleshooting-enable-disable-debug-output-api"></a>
+
+Every feature is just an Icinga config object similar to Host and Service, e.g.
```
Comment (yhabteab) [inline doc/15-troubleshooting.md:181]: `s/Icinga/Icinga 2/`.

File: doc/15-troubleshooting.md [hunk 2]
```diff
@@ -176,6 +176,60 @@ C:\> cd C:\ProgramData\icinga2\var\log\icinga2
...
+the `debuglog` feature is a [FileLogger](09-object-types.md#objecttype-filelogger).
+Those can also be [managed via API](12-icinga2-api.md#icinga2-api-config-objects)
+at runtime. This is a good alternative to `icinga2 feature enable debuglog`:
+
+* Object creation/deletion via API happens immediately and requires no restart
+* Hence, the debug log is enabled exactly as long as desired
+
+However, in case of [a HA zone](06-distributed-monitoring.md#distributed-monitoring-scenarios-ha-master-agents)
+the feature gets enabled/disabled on both nodes once they're connected.
```
Comment (yhabteab) [inline doc/15-troubleshooting.md:190]: Since this is like any other configuration object, is it worth mentioning some HA specific details? If in doubt, I would rephrase it slightly different, as I can currently think of it as both masters need to be waiting to each other to be connected in order to get enabled the `debuglog` on both sides.

**Review by yhabteab — COMMENTED** [review]
Pls. rebase!

File: doc/15-troubleshooting.md (hunk 2 above)
Comment (Al2Klimov) [inline doc/15-troubleshooting.md:190]: # ⚠️ Thing to consider

* #10022

Comment (Al2Klimov) [reply doc/15-troubleshooting.md:190]: > Since this is like any other configuration object, is it worth mentioning some HA specific details?

Given that feature's I/O impact, no. It's not as trivial as every other object.

> I can currently think of it as both masters need to be waiting to each other to be connected in order to get enabled the `debuglog` on both sides.

Exactly! As long as the other master isn't connected, it doesn't write the debuglog.

Comment (yhabteab) [reply doc/15-troubleshooting.md:190]: > Exactly! As long as the other master isn't connected, it doesn't write the debuglog.

Seriously? How? Why?

Comment (Al2Klimov) [reply doc/15-troubleshooting.md:190]: I'm sorry.

~As long as the other master isn't connected, it doesn't write the debuglog.~

As long as the other master isn't connected, it doesn't start writing the debuglog too as it didn't get the runtime debug logger yet.

Comment (yhabteab) [reply doc/15-troubleshooting.md:190]: > As long as the other master isn't connected, it doesn't start writing the debuglog too as it didn't get the runtime debug logger yet.

I'm not questioning that! I'm just confused, and maybe others will be too, by this one `... the feature gets enabled/disabled on both nodes once they're connected.` The one specific endpoint that receives the API request does not have to wait for anyone to enable/disable it.

File: doc/15-troubleshooting.md [hunk 3]
```diff
@@ -176,6 +176,62 @@ C:\> cd C:\ProgramData\icinga2\var\log\icinga2
...
 
+### Enable/Disable Debug Output on the fly <a id="troubleshooting-enable-disable-debug-output-api"></a>
+
+Every feature is just an Icinga 2 config object similar to Host and Service, e.g.
+the `debuglog` feature is a [FileLogger](09-object-types.md#objecttype-filelogger).
+Those can also be [managed via API](12-icinga2-api.md#icinga2-api-config-objects)
+at runtime. This is a good alternative to `icinga2 feature enable debuglog`:
+
+* Object creation/deletion via API happens immediately and requires no restart
+* Hence, the debug log is enabled exactly as long as desired
```
Comment (oxzi) [inline doc/15-troubleshooting.md:187]:
```suggestion
at runtime.

This is a good alternative to `icinga2 feature enable debuglog` as object
creation/deletion via API happens immediately and requires no restart.
```

I would just continue with text, as the list makes it harder to read. Furthermore, the second bullet point adds no real value, imo.

**Review by oxzi — COMMENTED** [review]
I like this feature and think it might help users if we document it. Thus, unless there are technical limitations, I would like to see it in the manual.

File: doc/15-troubleshooting.md (hunk 3 above)
Comment (Al2Klimov) [reply doc/15-troubleshooting.md:187]: > the second bullet point adds no real value, imo.

Well, if your reload takes 1/2h, your debug log is enabled for 1/2h longer than necessary, spamming the disk. And if your setup is that large, it will spam a lot.

Comment (oxzi) [reply doc/15-troubleshooting.md:187]: This totally makes sense. Could you then please add this information to the documentation, because for me this was not the obvious intent of the second sentence.

Btw, I am in favor of this PR and come back to it occasionally, for example, just this morning.

File: doc/15-troubleshooting.md [hunk 4]
```diff
@@ -176,6 +176,65 @@ C:\> cd C:\ProgramData\icinga2\var\log\icinga2
...
 ```
 
+### Enable/Disable Debug Output on the fly <a id="troubleshooting-enable-disable-debug-output-api"></a>
+
+Every feature is just an Icinga 2 config object similar to Host and Service, e.g.
+the `debuglog` feature is a [FileLogger](09-object-types.md#objecttype-filelogger).
+Those can also be [managed via API](12-icinga2-api.md#icinga2-api-config-objects)
+at runtime. This is a good alternative to `icinga2 feature enable debuglog`:
+
+* Object creation/deletion via API happens immediately and requires no restart
```
Comment (oxzi) [inline doc/15-troubleshooting.md:186]: This section starts with a very technical detail instead of a description of this feature. May I suggest another opening:

```suggestion
The `debuglog` feature can also be created and deleted at runtime without having to restart Icinga 2.
Technically, this is possible because this feature is a  [FileLogger](09-object-types.md#objecttype-filelogger) that can be managed through the [API](12-icinga2-api.md#icinga2-api-config-objects).

This is a good alternative to `icinga2 feature enable debuglog` as object
creation/deletion via API happens immediately and requires no restart.
```

File: doc/15-troubleshooting.md (hunk 3 above)
Comment (oxzi) [reply doc/15-troubleshooting.md:187]: As stated above, the bullet list is hard to read. The second point is obsolete with the just introduced paragraph below. Please remove it.

File: doc/15-troubleshooting.md [hunk 5]
```diff
@@ -176,6 +176,64 @@ C:\> cd C:\ProgramData\icinga2\var\log\icinga2
 C:\ProgramData\icinga2\var\log\icinga2> Get-Content .\debug.log -tail 10 -wait
 ```
 
+### Enable/Disable Debug Output on the fly <a id="troubleshooting-enable-disable-debug-output-api"></a>
+
+The `debuglog` feature can also be created and deleted at runtime without having to restart Icinga 2.
+Technically, this is possible because this feature is a  [FileLogger](09-object-types.md#objecttype-filelogger)
```
Comment (oxzi) [inline doc/15-troubleshooting.md:182]:
```suggestion
Technically, this is possible because this feature is a [FileLogger](09-object-types.md#objecttype-filelogger)
```

**Review by oxzi — COMMENTED** [review]
Just this small nitpick of a double space, otherwise I am happy with it.

File: doc/15-troubleshooting.md (hunk 5 above)
Comment (Al2Klimov) [reply doc/15-troubleshooting.md:182]: Cause: https://github.com/Icinga/icinga2/pull/9981#discussion_r1800589430 :-)

Comment (oxzi) [reply doc/15-troubleshooting.md]: Hehe :>

File: doc/15-troubleshooting.md [hunk 6]
```diff
@@ -176,6 +176,64 @@ C:\> cd C:\ProgramData\icinga2\var\log\icinga2
...
+
+This is a good alternative to `icinga2 feature enable debuglog` as object
+creation/deletion via API happens immediately and requires no restart.
+
+The above matters in setups large enough for the reload to take a while.
+Especially these produce a lot of debug log output until disabled again.
+
+!!! info
+
+    In case of [a HA zone](06-distributed-monitoring.md#distributed-monitoring-scenarios-ha-master-agents)
```
Comment (oxzi) [inline doc/15-troubleshooting.md:193]:
```suggestion
    In case of [an HA zone](06-distributed-monitoring.md#distributed-monitoring-scenarios-ha-master-agents),
```

Sorry, I have missed this one before. Please excuse the additional round :/

**Review by oxzi — APPROVED** [review]

File: doc/15-troubleshooting.md (hunk 2 above)
Comment (yhabteab) [reply doc/15-troubleshooting.md]: What did happen to this consideration? I mean it's clearly a bug, but I didn't get why posted such a prominent warning here and doing nothing about it.

Comment (Al2Klimov) [reply doc/15-troubleshooting.md]: > why posted such a prominent warning here

This PR (necessarily) "sells" the user a shiny HA feature which has a bug which may negatively affect what the user's doing inspired by this PR. Hence, you as reviewers may reconsider the PR as whole...

> and doing nothing about it

... because, should we really make a disclaimer in the docs like, HA may be not reliable?

Comment (yhabteab) [reply doc/15-troubleshooting.md]: > ... because, should we really make a disclaimer in the docs like, HA may be not reliable?

No, I wouldn't do it either, as the bug literally impacts every Icinga 2 object. I just didn't get the purpose of this comment with a prominent and eye catching warning sign.

**Review by yhabteab — APPROVED** [review]
